microcosm_fastapi/audit.py

```diff
@@ -0,0 +1,350 @@
+from microcosm_fastapi.errors import (
+    extract_context,
+    extract_error_message,
+    extract_include_stack_trace,
+    extract_status_code,
+)
+from microcosm_fastapi.utils import AsyncIteratorWrapper
+
+
+DEFAULT_INCLUDE_REQUEST_BODY = 400
+DEFAULT_INCLUDE_RESPONSE_BODY = 400
+ERROR_MESSAGE_LIMIT = 2048
+
+AUDIT_LOGGER_NAME = "audit"
+
+
+AuditOptions = namedtuple("AuditOptions", [
+    "include_request_body",
+    "include_response_body",
+    "include_path",
+    "include_query_string",
+    "log_as_debug",
+])
+
+
+SKIP_LOGGING = "_microcosm_flask_skip_audit_logging"
+
+
+def is_uuid(value):
+    try:
+        UUID(value)
+        return True
+    except Exception:
+        return False
+
+
+def should_skip_logging(request: Request):
+    """
+    Should we skip logging for this handler?
+
+    """
+    return strtobool(request.headers.get("x-request-nolog", "false"))
+
+
+class RequestInfo:
+    """
+    Capture of key information for requests.
+
+    """
+    def __init__(self, options: AuditOptions, request: Request, request_context: Dict[str, Any], app_metadata: Metadata):
+        self.options = options
+        self.app_metadata = app_metadata
+        self.operation = None
+        self.func = None
+        self.method = request.method
+        self.args = request.query_params
+
+        self.request = request
+        self.path = request.url.path
+        self.query = request.url.query
+        self.request_state = request.state
+
+        self.request_context = request_context
+        self.timing = dict()
+
+        self.error = None
+        self.stack_trace = None
+        self.request_body = None
+        self.response_body = None
+        self.response_headers = None
+        self.status_code = None
+        self.success = None
+
+    def to_dict(self):
+        dct = dict(
+            operation=self.operation,
+            func=self.func,
+            method=self.method,
+            **self.timing
+        )
+
+        if self.options.include_query_string and self.args:
+            dct.update({
+                key: values
+                for key, values in self.args.items()
+            })
+
+        if self.request_context is not None:
+            dct.update(self.request_context)
+
+        if self.success is True:
+            dct.update(
+                success=self.success,
+                status_code=self.status_code,
+            )
+        if self.success is False:
+            dct.update(
+                success=self.success,
+                message=extract_error_message(self.error)[:ERROR_MESSAGE_LIMIT],
+                context=extract_context(self.error),
+                stack_trace=self.stack_trace,
+                status_code=self.status_code,
+            )
+
+        self.post_process_request_body(dct)
+        self.post_process_response_body(dct)
+        self.post_process_response_headers(dct)
+
+        return dct
+
+    def log(self, logger):
```

Comment:
It seems this class has too many responsibilities. I'd prefer to see this within middleware.

Reply:
Fair enough - have moved out into the middleware 👍